src/main/java/org/joshsim/util/WireResponseParser.java

```diff
@@ -0,0 +1,235 @@
+      this.errorMessage = null;
+    }
+
+    /**
+     * Constructor for END responses.
+     *
+     * @param replicateNumber The replicate number that completed
+     */
+    public ParsedResponse(ResponseType type, int replicateNumber) {
+      if (type != ResponseType.END) {
+        throw new IllegalArgumentException("This constructor is only for END responses");
+      }
+      this.type = type;
+      this.replicateNumber = replicateNumber;
+      this.dataLine = null;
+      this.stepCount = 0;
+      this.errorMessage = null;
+    }
+
+    /**
+     * Constructor for ERROR responses.
+     *
+     * @param errorMessage The error message from the remote server
+     */
+    public ParsedResponse(String errorMessage) {
+      this.type = ResponseType.ERROR;
+      this.replicateNumber = -1;
+      this.dataLine = null;
+      this.stepCount = 0;
+      this.errorMessage = errorMessage;
+    }
+
+    /**
+     * Gets the response type.
+     *
+     * @return The type of this response
+     */
+    public ResponseType getType() {
+      return type;
+    }
+
+    /**
+     * Gets the replicate number.
+     *
+     * @return The replicate number, or -1 if not applicable
+     */
+    public int getReplicateNumber() {
+      return replicateNumber;
+    }
+
+    /**
+     * Gets the data line for DATUM responses.
+     *
+     * @return The wire format data line, or null if not a DATUM response
+     */
+    public String getDataLine() {
+      return dataLine;
+    }
+
+    /**
+     * Gets the step count for PROGRESS responses.
+     *
+     * @return The step count, or 0 if not a PROGRESS response
+     */
+    public long getStepCount() {
+      return stepCount;
+    }
+
+    /**
+     * Gets the error message for ERROR responses.
+     *
+     * @return The error message, or null if not an ERROR response
+     */
+    public String getErrorMessage() {
+      return errorMessage;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("ParsedResponse{type=%s, replicate=%d, step=%d, data='%s', error='%s'}",
+          type, replicateNumber, stepCount, dataLine, errorMessage);
+    }
+  }
+
+  /**
+   * Parses a single response line from the remote engine.
+   *
+   * <p>This method analyzes a line from the streaming HTTP response and determines
+   * its type and content. It supports the following formats:</p>
+   * <ul>
+   *   <li>[end N] - Replicate N has completed</li>
+   *   <li>[progress N] - Current step is N</li>
+   *   <li>[error] message - An error occurred with the given message</li>
+   *   <li>[N] data - Data point from replicate N</li>
+   *   <li>[N] - Empty data point from replicate N (ignored)</li>
+   * </ul>
+   *
+   * @param line The response line to parse
+   * @return A ParsedResponse object, or null if the line should be ignored
+   * @throws IllegalArgumentException if the line format is invalid
+   */
+  public static ParsedResponse parseEngineResponse(String line) {
```

Comment:
Let's break this into a series of `WireResponseParserStrategy` implementers. Let's have one per matcher we have here. Let's have an `Optional<WireParseResult>` returned by each to avoid returning null. Let's have it be empty if it doesn't apply. Let's make a list of these and apply until one returns a non-empty response. We can have a flag for the line to be ignored. We can change the return value of `parseEngineResponse` to be `Optional<ParsedResponse>`